### PR TITLE
Implement a baseline algorithm

### PR DESCRIPTION
**Pull request by Thanoshan**
This is a work-in-progress PR for implementing a baseline algorithm.

Currently, this PR does the following:
- [x] Fix's @weihangzheng's implementation of communicated messages to only add noise for messages that are sent by an adversarial agent, and to return communciation messages dictionary properly
- [x] Removes "beliefs" from "Infos" since we'll be passing the instance of an Agent to the algorithm. Infos cannot be modified by the algorithm and only provides read-only data, and we need the algorithm to be able to update agent beliefs.
- [x] Implements baseline algorithm converted to a discrete space. We strip the extreme D values from the communications and average the rest to compute our position delta for any belief 
- [x] Implements a slightly modified version of baseline that deviates from the local optimal action with a probability epsilon
- [x] Added a marker for highlighting the target on the grid

### CODE REVIEW

**Review by pandyah5 — COMMENTED** [review]
LGTM, great work! I have left minor comments to address

File: src/nepiada/baseline.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            continue
+
+        in_messages = []
+
+        # Get incoming messages that contain this agent's position
+        for _, comm_message in incoming_messages[current_agent].items():
+            if comm_message is not None:
+                in_messages.append(comm_message)
+
+        # Strip the extreme values
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/baseline.py]: Done.

File: src/nepiada/baseline.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+def main():
+    env_config = BaselineConfig()
+
+    env = nepiada.parallel_env(config=env_config)
+    observations, infos = env.reset()
+
+    while env.agents:
+        actions = {}
+        for c_agent in env.agents:
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/baseline.py]: Fixed, good catch.

File: src/nepiada/baseline.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        y_pos_delta = sum(y_pos_deviation) / len(y_pos_deviation)
+
+        # Update the beliefs
+        new_beliefs[current_agent] = (
+            curr_beliefs[current_agent][0] + x_pos_delta,
+            curr_beliefs[current_agent][1] + y_pos_delta,
+        )
+
+
+def step(agent_name, agent_instance, observations, infos, env, config):
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/baseline.py:140]: Good catch. fixed.

File: src/nepiada/baseline.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+            # What do we do here?
+            continue
+        target_neighbor_cost += np.sqrt(
+            (curr_agent_position[0] - agent_position[0] - target_relative_position[0])
+            ** 2
+            + (curr_agent_position[1] - agent_position[1] - target_relative_position[1])
+            ** 2
+        )
+
+    # Return cost, should this be weighted?
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/baseline.py]: Good catch, fixed.

File: src/nepiada/baseline.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+    # Calculate the target neighbour cost
+    for curr_agent_name, target_relative_position in target_neighbours.items():
+        assert beliefs[agent_name] is not None
+
+        curr_agent_position = beliefs[curr_agent_name]
+        agent_position = beliefs[agent_name]
+
+        if curr_agent_position is None:
+            # What do we do here?
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/baseline.py]: Done.

File: src/nepiada/env/nepiada.py [hunk 6]
```diff
@@ -79,7 +80,7 @@ def observation_space(self, agent):
     @functools.lru_cache(maxsize=None)
     def action_space(self, agent):
         # Discrete movement, either up, down, stay, left or right.
-        return Discrete(5)
+        return Discrete(len(self.config.possible_moves))
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/env/nepiada.py:83]: Thank you! :D

File: src/nepiada/env/nepiada.py [hunk 7]
```diff
@@ -123,38 +124,52 @@ def get_observations(self):
...
 
             for target_agent_name in self.agents:
-
                 incoming_communcation_messages = {}
 
-                if not observation[target_agent_name]: # Must estimate where the agent is via communication
-                    for helpful_agent in self.world.graph.comm[agent_name]:  
-                        helpful_beliefs = self.config.noise.add_noise(self.world.get_agent(helpful_agent).beliefs)
+                if not observation[
+                    target_agent_name
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/env/nepiada.py]: The Ruff formatter insists this is better formatting. I'll revert it, but it'll probably happen again in the future. I don't think its that big of a deal

File: src/nepiada/env/nepiada.py [hunk 8]
```diff
@@ -245,16 +266,18 @@ def step(self, actions):
...
-                        
+
         # Info will be used to pass information about comm graphs, beliefs, and incoming messages
         self.infos = {agent_name: {} for agent_name in self.agents}
         for agent_name in self.agents:
             self.infos[agent_name]["comm"] = self.world.graph.comm[agent_name]
-            self.infos[agent_name]["beliefs"] = self.world.get_agent(agent_name).beliefs
-            self.infos[agent_name]["incoming_messages"] = incoming_all_messages[agent_name]
+            self.infos[agent_name]["incoming_messages"] = incoming_all_messages[
+                agent_name
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/env/nepiada.py:277]: Ruff insists on this one because it exceeds the max line length, which I agree with.

**Review by Thanoshan — COMMENTED** [review]
Thanks Hetav!

**Review by pandyah5 — APPROVED** [review]
LGTM

**Pull request merged**